ray-operator/controllers/ray/networkpolicy_controller.go

```diff
@@ -0,0 +1,460 @@
+			From: []networkingv1.NetworkPolicyPeer{*rayJobPeer},
+		})
+	}
+
+	return &networkingv1.NetworkPolicy{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-head", instance.Name),
+			Namespace: instance.Namespace,
+			Labels:    labels,
+		},
+		Spec: networkingv1.NetworkPolicySpec{
+			PodSelector: metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					utils.RayClusterLabelKey:  instance.Name,
+					utils.RayNodeTypeLabelKey: string(rayv1.HeadNode),
+				},
+			},
+			PolicyTypes: []networkingv1.PolicyType{networkingv1.PolicyTypeIngress},
+			Ingress:     ingressRules,
+		},
+	}
+}
+
+// buildWorkerNetworkPolicy creates a NetworkPolicy for Ray worker pods
+func (r *NetworkPolicyController) buildWorkerNetworkPolicy(instance *rayv1.RayCluster) *networkingv1.NetworkPolicy {
+	labels := map[string]string{
+		utils.RayClusterLabelKey:                instance.Name,
+		utils.KubernetesApplicationNameLabelKey: utils.ApplicationName,
+		utils.KubernetesCreatedByLabelKey:       utils.ComponentName,
+	}
+
+	return &networkingv1.NetworkPolicy{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-workers", instance.Name),
+			Namespace: instance.Namespace,
+			Labels:    labels,
+		},
+		Spec: networkingv1.NetworkPolicySpec{
+			PodSelector: metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					utils.RayClusterLabelKey:  instance.Name,
+					utils.RayNodeTypeLabelKey: string(rayv1.WorkerNode),
+				},
+			},
+			PolicyTypes: []networkingv1.PolicyType{networkingv1.PolicyTypeIngress},
+			Ingress: []networkingv1.NetworkPolicyIngressRule{
+				{
+					From: []networkingv1.NetworkPolicyPeer{
+						{
+							PodSelector: &metav1.LabelSelector{
+								MatchLabels: map[string]string{
+									utils.RayClusterLabelKey: instance.Name,
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+// buildRayJobPeer creates a NetworkPolicy peer for RayJob submitter pods
+// Returns nil if RayCluster is not owned by RayJob
+func (r *NetworkPolicyController) buildRayJobPeer(instance *rayv1.RayCluster) *networkingv1.NetworkPolicyPeer {
+	// Check if RayCluster is owned by RayJob
+	for _, ownerRef := range instance.OwnerReferences {
+		if ownerRef.Kind == "RayJob" {
+			// Return peer for RayJob submitter pods
+			return &networkingv1.NetworkPolicyPeer{
+				PodSelector: &metav1.LabelSelector{
+					MatchLabels: map[string]string{
+						"batch.kubernetes.io/job-name": ownerRef.Name,
+					},
+				},
+			}
+		}
+	}
+	// No RayJob owner = no RayJob submitter pods to allow
+	return nil
+}
+
+// isMTLSEnabled checks if mTLS is enabled for the RayCluster
+// This looks for TLS-related environment variables or configuration
+func (r *NetworkPolicyController) isMTLSEnabled(instance *rayv1.RayCluster) bool {
+	// Check head group for TLS environment variables
+	if r.checkContainersForMTLS(instance.Spec.HeadGroupSpec.Template.Spec.Containers) {
```

Comment:
these checks wouldn't be valid anymore given that we configure it directly on the resources